src/dmd/doc.d

```diff
@@ -1948,6 +2007,201 @@ Lno:
+                    // invalid link
+                    href.length = 0;
+                    goto LEndInlineLink;
+                }
+                if (!href.length)
+                    href = (cast(const char*) buf.data)[iHrefStart .. j];
+                break;
+            case '\r':
+            case '\n':
+                if (iLinkNewline == j || !href.length)
+                {
+                    // no blank lines or newlines in hrefs
+                    href.length = 0;
+                    goto LEndInlineLink;
+                }
+                iLinkNewline = j + 1;
+                break;
+            default:
+                break;
+            }
+        }
+    LEndInlineLink:
+        if (href.length)
+        {
+            size_t iAfterLink = i - delimiter.count;
+            string macroName;
+            href = href.dup;
+            if (title.length)
+            {
+                title = title.dup;
+                if (delimiter.type == '[')
+                    macroName = "$(LINK_TITLE ";
+                else
+                    macroName = "$(IMAGE_TITLE ";
+            }
+            else
+            {
+                if (delimiter.type == '[')
+                    macroName = "$(LINK2 ";
+                else
+                    macroName = "$(IMAGE ";
+            }
+            buf.remove(delimiter.iStart, delimiter.count);
+            buf.remove(i - 1, j - i + 1);
+            j = buf.insert(delimiter.iStart, macroName);
+            j = buf.insert(j, href);
+            j = buf.insert(j, ", ");
+            iAfterLink += macroName.length + href.length + 2;
+            if (title.length)
+            {
+                j = buf.insert(j, title);
+                j = buf.insert(j, ", ");
+                iAfterLink += title.length + 2;
+            }
+            buf.insert(iAfterLink, ")");
+            return iAfterLink;
+        }
+    }
+    return i;
+}
+
+/****************************************************
+ * Replace a Markdown thematic break (HR).
+ */
+bool replaceMarkdownThematicBreak(OutBuffer *buf, ref size_t i, size_t iLineStart)
+{
+    const slice = buf.peekSlice();
+    char c = buf.data[i];
+    size_t j = i + 1;
+    int repeat = 1;
+    for (; j < slice.length; j++)
+    {
+        if (buf.data[j] == c)
+            ++repeat;
+        else if (buf.data[j] != ' ' && buf.data[j] != '\t')
+            break;
+    }
+    if (repeat >= 3)
+    {
+        if (j >= buf.offset || buf.data[j] == '\n' || buf.data[j] == '\r')
+        {
+            buf.remove(iLineStart, j - iLineStart);
+            i = buf.insert(iLineStart, "$(HR)") - 1;
+            return true;
+        }
+    }
+    return false;
+}
+
+/****************************************************
+ * End a Markdown heading, if inside one.
+ */
+bool endMarkdownHeading(OutBuffer *buf, ref size_t i, ref int headingLevel, size_t iHeadingStart)
+{
+    if (headingLevel)
+    {
+        char* heading = cast(char*) "$(H0 ".dup;
```

Comment:
`.ptr` will allow you to avoid the cast

Reply:
or make it an array. Then line 2192 is safe, and use .ptr in the only other use, line 2193.

Reply:
On further review, there is no need for `heading` to be allocated on the heap. It's a short, fixed size. Make it a static array on the stack.